packages/contracts/contracts/LenderCommitmentForwarder/extensions/CommitmentRolloverLoan.sol

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+// Contracts
+import "@openzeppelin/contracts-upgradeable/utils/AddressUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+
+// Interfaces
+import "../../interfaces/ITellerV2.sol";
+import "../../interfaces/IProtocolFee.sol";
+import "../../interfaces/ITellerV2Storage.sol";
+import "../../interfaces/IMarketRegistry.sol";
+import "../../interfaces/ILenderCommitmentForwarder.sol";
+import "../../interfaces/ICommitmentRolloverLoan.sol";
+import "../../libraries/NumbersLib.sol";
+
+contract CommitmentRolloverLoan is ICommitmentRolloverLoan {
+    using AddressUpgradeable for address;
+    using NumbersLib for uint256;
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    ITellerV2 public immutable TELLER_V2;
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    ILenderCommitmentForwarder public immutable LENDER_COMMITMENT_FORWARDER;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(address _tellerV2, address _lenderCommitmentForwarder) {
+        TELLER_V2 = ITellerV2(_tellerV2);
+        LENDER_COMMITMENT_FORWARDER = ILenderCommitmentForwarder(
+            _lenderCommitmentForwarder
+        );
+    }
+
+    /**
+     * @notice Allows a borrower to rollover a loan to a new commitment.
+     * @param _loanId The ID of the existing loan.
+     * @param _rolloverAmount The amount to rollover.
+     * @param _commitmentArgs Arguments for the commitment to accept.
+     * @return newLoanId_ The ID of the new loan created by accepting the commitment.
+     */
+    function rolloverLoan(
+        uint256 _loanId,
+        uint256 _rolloverAmount,
+        AcceptCommitmentArgs calldata _commitmentArgs
+    ) external returns (uint256 newLoanId_) {
+        address borrower = TELLER_V2.getLoanBorrower(_loanId);
+        require(borrower == msg.sender, "CommitmentRolloverLoan: not borrower");
+
+        // Get lending token and balance before
+        IERC20Upgradeable lendingToken = IERC20Upgradeable(
+            TELLER_V2.getLoanLendingToken(_loanId)
+        );
+        uint256 balanceBefore = lendingToken.balanceOf(address(this));
+
+        if (_rolloverAmount > 0) {
+            //accept funds from the borrower to this contract
+            lendingToken.transferFrom(borrower, address(this), _rolloverAmount);
+        }
+
+        // Accept commitment and receive funds to this contract
+        newLoanId_ = _acceptCommitment(_commitmentArgs);
+
+        // Calculate funds received
+        uint256 fundsReceived = lendingToken.balanceOf(address(this)) -
+            balanceBefore;
```

Comment:
what if we simplified `fundsReceived` to just 
```solidity
uint256 fundsRecieved = _rolloverAmount + _commitmentArgs.principalAmount;
```
removing need to make extra `balanceOf` call